### PR TITLE
Create stack for RAND/R&D components

### PR DESCRIPTION
**Pull request by JoelClemence**
Creates:
* Domain (`rand.aws.chdev.org`)
* SSL Certificate
* Application Load Balancer (using internal Terraform module)
* ECS Cluster
* SSM Parameter(s) pertinent to resources provisioned

### CODE REVIEW

File: groups/infrastructure/main.tf [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+  type    = aws_acm_certificate.certificate.domain_validation_options[0].resource_record_type
+  zone_id = data.aws_route53_zone.hosted_zone.id
+  records = [aws_acm_certificate.certificate.domain_validation_options[0].resource_record_value]
+  ttl     = 60
+}
+
+resource "aws_acm_certificate_validation" "certificate" {
+  certificate_arn         = aws_acm_certificate.certificate.arn
+  validation_record_fqdns = [aws_route53_record.certificate_validation.fqdn]
+}
```
Comment (sienkin) [inline groups/infrastructure/main.tf]: We shouldn't have resources being defined in the `main.tf` alongside terraform, provider and module config.

Consider splitting these out in to appropriate files.

File: groups/infrastructure/main.tf [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+  name = "/${local.name_prefix}/alb_name"
+  type = "SecureString"
+  value = module.alb.application_load_balancer_name
+}
+
+resource "aws_ssm_parameter" "rand_alb_sg" {
+  name = "/${local.name_prefix}/alb_security_group_id"
+  type = "SecureString"
+  value = module.alb.security_group_id
+}
```
Comment (sienkin) [inline groups/infrastructure/main.tf]: See https://github.com/companieshouse/rand-stack/pull/1/files#r1420058007

File: groups/infrastructure/data.tf [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+  }
+
+  tags = {
+    NetworkType = "private"
+  }
+}
+
+data "aws_prefix_list" "admin" {
+  name = local.admin_prefix_list_name
+}
```
Comment (sienkin) [inline groups/infrastructure/data.tf]: This isn't the ideal resource to use for this lookup. You should use `aws_ec2_managed_prefix_list`.

File: groups/infrastructure/data.tf [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+
+data "aws_subnets" "private" {
+  filter {
+    name   = "vpc-id"
+    values = [data.aws_vpc.vpc.id]
+  }
+
+  tags = {
+    NetworkType = "private"
+  }
```
Comment (sienkin) [inline groups/infrastructure/data.tf]: This filter is insufficient and instead we should configure a second filter that performs a lookup based on a pattern. Such as in the style of `subnet-name-*`

This pattern would be held in Vault and looked-up accordingly.

E.g.:
```
  filter {
    name   = "vpc-id"
    values = [data.aws_vpc.vpc.id]
  }

  filter{
    name   = "tag:Name"
    values = [local.subnet_pattern]
  }
```

File: groups/infrastructure/main.tf [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+
+  environment         = var.environment
+  service             = local.service_name
+  ssl_certificate_arn = aws_acm_certificate.certificate.arn
+  subnet_ids          = data.aws_subnets.private.subnet_ids
+  vpc_id              = data.aws_vpc.vpc.id
+
+  create_security_group  = true
+  
+  ingress_prefix_lists   = local.asg_ingress_prefix_list
```
Comment (sienkin) [inline groups/infrastructure/main.tf]: The module doesn't define this input - it should be `ingress_prefix_list_ids`

You should review the modules you're referencing to ensure you're satisfying their requirements.

File: groups/infrastructure/main.tf (hunk 2 above)
Comment (sienkin) [inline groups/infrastructure/main.tf]: We're not specifying a KMS key for these and so the default will be used. Is this adequate or should we be creating a KMS key for this?

Also I don't understand why we're creating parameters for these values - they look like they're meant to be Terraform outputs.

**Comment by sienkin** [comment]
The format of this code needs attention. Consider running `terraform fmt` or similar to ensure the code follows Hashicorp's [Style Convention](https://developer.hashicorp.com/terraform/language/syntax/style)

There's also inconsistencies in the use of `rand` and `randd` which should be looked at.

File: groups/infrastructure/main.tf (hunk 2 above)
Comment (JoelClemence) [reply groups/infrastructure/main.tf]: I'll set a KMS key.

They will be referenced by other Terraform deployments therefore using terraform outputs would mean reading the state file into the downstream Terraform deployments which I was a little unsure about as I seem to remember reading this was not a pattern which should be done any more. Using Parameters to allow downstream deployments to reference these as SSM parameters. Unless I am missing something with `output`s?

**Comment by JoelClemence** [comment]
> The format of this code needs attention. Consider running `terraform fmt` or similar to ensure the code follows Hashicorp's [Style Convention](https://developer.hashicorp.com/terraform/language/syntax/style)
> 
> There's also inconsistencies in the use of `rand` and `randd` which should be looked at.

Formatted using `terraform fmt`

Regarding `rand` so, this is up for discussion, technically speaking the name of my team is `rand` which is why domain etc is `rand`, my thoughts behind the environment name being `randd` is that `rand` may not be the only team doing R&D so using a common environment namespace may be ok, but happy to go for `rand` only if deemed unnecessary/not the appropriate direction.

File: groups/infrastructure/alb.tf [hunk 6]
```diff
@@ -0,0 +1,26 @@
+module "alb" {
+  source = "git@github.com:companieshouse/terraform-modules//aws/application_load_balancer?ref=1.0.205"
+
+  environment         = var.environment
+  service             = local.service_name
+  ssl_certificate_arn = aws_acm_certificate.certificate.arn
+  subnet_ids          = data.aws_subnets.private.subnet_ids
```
Comment (sienkin) [inline groups/infrastructure/alb.tf]: `subnet_ids` isn't an exported attribute for this data source. This should be `ids`

File: groups/infrastructure/route-53.tf [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+  type    = aws_acm_certificate.certificate.domain_validation_options[0].resource_record_type
+  zone_id = data.aws_route53_zone.hosted_zone.id
+  records = [aws_acm_certificate.certificate.domain_validation_options[0].resource_record_value]
+  ttl     = 60
+}
+
+resource "aws_acm_certificate_validation" "certificate" {
+  certificate_arn         = aws_acm_certificate.certificate.arn
+  validation_record_fqdns = [aws_route53_record.certificate_validation.fqdn]
+}
```
Comment (sienkin) [inline groups/infrastructure/route-53.tf]: You're referencing a resource `aws_route53_record` with label `certificate_validation` but one doesn't exist. You have one with a label of `route` and so either the label on the `aws_route53_record` resource should be updated accordingly, or the reference updated to `aws_route53_record.route`

File: groups/infrastructure/route-53.tf [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+  }
+}
+
+resource "aws_route53_record" "route" {
+  name    = aws_acm_certificate.certificate.domain_validation_options[0].resource_record_name
+  type    = aws_acm_certificate.certificate.domain_validation_options[0].resource_record_type
+  zone_id = data.aws_route53_zone.hosted_zone.id
+  records = [aws_acm_certificate.certificate.domain_validation_options[0].resource_record_value]
+  ttl     = 60
+}
```
Comment (sienkin) [inline groups/infrastructure/route-53.tf]: This is legacy syntax and is not valid with the version of the `aws` provider defined in the code. `domain_validation_options` is now exported as a set of objects and so must be iterated over via a `for_each` loop.

The [provider documentation](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/acm_certificate_validation) has a complete example which will also require a small update to the `aws_acm_certificate_validation` resource to fit in with the updated syntax.

**Review by sienkin — APPROVED** [review]

**Review by ch-tstrickland — APPROVED** [review]

**Pull request merged**